[quant][pyper] Add embedding_bag weight quantize and dequantize ops

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#41293 [quant][pyper] Add embedding_bag weight quantize and dequantize ops**

Summary:

Add new operators that does quantize and packing for 8 bit and 4 bit embedding bag operators.
This is an initial change to help unblock testing. This will be follwed by adding graph mode passes to enable quantization of embedding_bag module

Note to reviewers: Future PRs will replace this op with a separate quantize and pack operator and add support for floating point scale and zero point.

Test Plan:
python test/test_quantization.py TestQuantizedEmbeddingBag

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22506700](https://our.internmc.facebook.com/intern/diff/D22506700)

supriyar committed:
[quant][pyper] Add embedding_bag weight quantize and dequantize ops
Summary:

Add new operators that does quantize and packing for 8 bit and 4 bit embedding bag operators.
This is an initial change to help unblock testing. This will be follwed by adding graph mode passes to enable quantization of embedding_bag module

Note to reviewers: Future PRs will replace this op with a separate quantize and pack operator and add support for floating point scale and zero point.

Test Plan:
python test/test_quantization.py TestQuantizedEmbeddingBag

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][pyper] Add embedding_bag weight quantize and dequantize ops"


Summary:

Add new operators that does quantize and packing for 8 bit and 4 bit embedding bag operators.
This is an initial change to help unblock testing. This will be follwed by adding graph mode passes to enable quantization of embedding_bag module

Note to reviewers: Future PRs will replace this op with a separate quantize and pack operator and add support for floating point scale and zero point.

Test Plan:
python test/test_quantization.py TestQuantizedEmbeddingBag

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

radkris-git commented on aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp:
```diff
@@ -0,0 +1,138 @@
+  constexpr int NUM_ELEM_PER_BYTE = 8 / BIT_RATE;
+  TORCH_CHECK(
+      weight_contig.size(weight.dim() - 1) % NUM_ELEM_PER_BYTE == 0,
+      "FloatToFused4BitRowwiseQuantizedOp only works for the number of "
```
I think this is a nit.

supriyar replied:
You mean this check isn't required?

radkris-git replied:
The error message says "FloatToFused4BitRowwiseQuantizedOp". :) It should be "qembeddingbag_4bit_prepack only works for the number of columns a multiple of 2".

radkris-git commented on test/quantization/test_quantized_op.py:
```diff
@@ -2703,23 +2703,77 @@ def test_qlinear_unpack(self, W, use_channelwise):
+            workspace.ResetWorkspace()
+
+            workspace.FeedBlob("weights", weights)
+            workspace.RunOperatorOnce(
```
Using C2 operators gives confidence that PT ops work. Do you plan to keep the comparison with C2 for the longer term as well?

supriyar replied:
No, we will add reference implementations eventually once C2 is deprecated.

supriyar committed:
Update on "[quant][pyper] Add embedding_bag weight quantize and dequantize ops"


Summary:

Add new operators that does quantize and packing for 8 bit and 4 bit embedding bag operators.
This is an initial change to help unblock testing. This will be follwed by adding graph mode passes to enable quantization of embedding_bag module

Note to reviewers: Future PRs will replace this op with a separate quantize and pack operator and add support for floating point scale and zero point.

Test Plan:
python test/test_quantization.py TestQuantizedEmbeddingBag

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22506700](https://our.internmc.facebook.com/intern/diff/D22506700)

[ghstack-poisoned]

vkuzo commented on aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp:
```diff
@@ -0,0 +1,138 @@
+  auto* output_data = output.data_ptr<uint8_t>();
+  const auto output_columns = output.size(output.dim() - 1);
+
+  for (int row = 0; row < embedding_rows; ++row) {
```
does performance matter, or is this a reference implementation?  Could probably parallelize if needed (same for the other op)

supriyar replied:
This op is a temporary solution until we de-couple quantize and packing. We can re-visit optimizations then.

vkuzo replied:
makes sense

vkuzo commented on aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp:
```diff
@@ -0,0 +1,138 @@
+        output_row[col / NUM_ELEM_PER_BYTE] |=
+            (quantized << ((col % NUM_ELEM_PER_BYTE) * BIT_RATE));
+      }
+    }
```
optional nit: `} // embedding_cols` for readability?

vkuzo commented on aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp:
```diff
@@ -0,0 +1,138 @@
+            (quantized << ((col % NUM_ELEM_PER_BYTE) * BIT_RATE));
+      }
+    }
+  }
```
optional nit: `} // embedding_rows` for readability?

vkuzo commented on aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp:
```diff
@@ -0,0 +1,138 @@
+    const float* input_row = weight_data + row * embedding_cols;
+    std::uint8_t* output_row = output_data + row * output_columns;
+
+    at::Half* output_row_scale_zp = reinterpret_cast<at::Half*>(
```
optional readability nit: if this is packed at the end of a row, maybe we can move the code down to be below the weight packing, so the code structure follows the data format?

vkuzo commented on aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp:
```diff
@@ -0,0 +1,138 @@
+          0,
+          std::min<int>(
+              lrintf((X - Xmin) * inverse_scale), (1 << BIT_RATE) - 1));
+      if (col % NUM_ELEM_PER_BYTE == 0) {
```
maybe we can add a quick comment on the data format? i.e. index 0 is packed in the low bits and index 1 in the high bits, etc

vkuzo commented on aten/src/ATen/native/quantized/cpu/qembeddingbag_unpack.cpp:
```diff
@@ -0,0 +1,85 @@
+Tensor qembeddingbag_byte_unpack(const Tensor& packed_weight) {
+  const auto input_rows = packed_weight.size(0);
+  const auto input_columns = packed_weight.size(1);
+  int output_columns = input_columns - 2 * sizeof(float);
```
nit: add comment that the 2 is coming from scale+zp?

vkuzo reviewed (commented):
lg, feel free to ignore the comments if this implementation will be replaced in the near future

supriyar committed:
Update on "[quant][pyper] Add embedding_bag weight quantize and dequantize ops"


Summary:

Add new operators that does quantize and packing for 8 bit and 4 bit embedding bag operators.
This is an initial change to help unblock testing. This will be follwed by adding graph mode passes to enable quantization of embedding_bag module

Note to reviewers: Future PRs will replace this op with a separate quantize and pack operator and add support for floating point scale and zero point.

Test Plan:
python test/test_quantization.py TestQuantizedEmbeddingBag

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22506700](https://our.internmc.facebook.com/intern/diff/D22506700)

[ghstack-poisoned]

supriyar committed:
Update on "[quant][pyper] Add embedding_bag weight quantize and dequantize ops"


Summary:

Add new operators that does quantize and packing for 8 bit and 4 bit embedding bag operators.
This is an initial change to help unblock testing. This will be follwed by adding graph mode passes to enable quantization of embedding_bag module

Note to reviewers: Future PRs will replace this op with a separate quantize and pack operator and add support for floating point scale and zero point.

Test Plan:
python test/test_quantization.py TestQuantizedEmbeddingBag

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22506700](https://our.internmc.facebook.com/intern/diff/D22506700)

[ghstack-poisoned]
